### PR TITLE
feat(gha): add a tagged container with the hash+timestamp

### PR DESCRIPTION
**Pull request by emilweth**
This PR add a tag for the `main` branch build, that allow to track new image every time something is merged in main

The format is `ghcr.io/mastodon/mastodon:edge-commit-hash-timestamp`

For exemple, with Flux you can add an [ImageRepository](https://fluxcd.io/flux/components/image/imagerepositories/) that points to `ghcr.io/mastodon/mastodon` and setup an [ImagePolicy](https://fluxcd.io/flux/components/image/imagepolicies/) like this : 

```YAML
apiVersion: image.toolkit.fluxcd.io/v1beta1
kind: ImagePolicy
metadata:
  name: mastodon-edge
  namespace: default
spec:
  imageRepositoryRef:
    name: mastodon-edge
  filterTags:
    pattern: '^edge-[a-f0-9]+-(?P<ts>[0-9]+)'
    extract: '$ts'
  policy:
    numerical:
      order: asc
```

And combined with an [ImageUpdateAutomation](https://fluxcd.io/flux/components/image/imageupdateautomations/), every time there's a new `:edge-xxx` tag, Flux will update the GitOps repo with this latest tag, and every deployments will be updated

### CODE REVIEW

**Comment by forsamori** [comment]
This isn't something I've come across before so I'd like to ask what the benefit of this is? Looking to learn more in this area so I thought I'd ask :)

**Comment by emilweth** [comment]
> This isn't something I've come across before so I'd like to ask what the benefit of this is? Looking to learn more in this area so I thought I'd ask :)

@forsamori on [my instance](https://rivals.space/) I use that with nightly builds, let me explain : 

I'm using [Flux](https://fluxcd.io/), to describe the entire state of my Kubernetes cluster using git.

I created an [ImageRepository](https://fluxcd.io/flux/components/image/imagerepositories/) resource that point to the mastodon docker repository, and I defined an [ImagePolicy](https://fluxcd.io/flux/components/image/imagepolicies/) that will take the latest image, with the `nightly-.*` pattern, like that : 

```YAML
apiVersion: image.toolkit.fluxcd.io/v1beta2
kind: ImageRepository
metadata:
  name: mastodon
  namespace: rivals-space-prod
spec:
  image: ghcr.io/mastodon/mastodon
  interval: 30m
  provider: generic
---
apiVersion: image.toolkit.fluxcd.io/v1beta2
kind: ImagePolicy
metadata:
  name: mastodon-nightly
  namespace: rivals-space-prod
spec:
  imageRepositoryRef:
    name: mastodon
  filterTags:
    pattern: 'nightly-(?P<y>\d{4})-(?P<m>\d{2})-(?P<d>\d{2})'
    extract: '$y$m$d'
  policy:
    numerical:
      order: asc
```

With this, flux will monitor the mastodon docker repository, and select a tag according to the ImagePolicy.

And using an [ImageUpdateAutomation](https://fluxcd.io/flux/components/image/imageupdateautomations/), Flux will automatically update the GitOps repository to the latest selected tag

```YAML
apiVersion: image.toolkit.fluxcd.io/v1beta1
kind: ImageUpdateAutomation
metadata:
  name: mastodon
  namespace: rivals-space-prod
spec:
  interval: 5m
  sourceRef:
    kind: GitRepository
    name: flux-apps
    namespace: flux-system
  git:
    commit:
      author:
        email: sloth@users.noreply.github.com
        name: Effective Sloth Bot
      messageTemplate: |
        feat: Bump container version
        
        Automation name: {{ .AutomationObject }}
        
        Files:
        {{ range $filename, $_ := .Updated.Files -}}
        - {{ $filename }}
        {{ end -}}

        Objects:
        {{ range $resource, $_ := .Updated.Objects -}}
        - {{ $resource.Kind }} {{ $resource.Name }}
        {{ end -}}

        Images:
        {{ range .Updated.Images -}}
        - {{.}}
        {{ end -}}
    push:
      branch: main
  update:
    path: ./rivals-space-prod
    strategy: Setters
```

```YAML
# mastodon-helm-values.yaml
images:
  mastodon:
    repository: "ghcr.io/mastodon/mastodon"
    pullPolicy: IfNotPresent
    tag: "nightly-2023-07-04" # {"$imagepolicy": "rivals-space-prod:mastodon-nightly:tag"}
 ```
 
 See the `# {"$imagepolicy": "rivals-space-prod:mastodon-nightly:tag"}` comment, this is what flux is looking for, and it will update the line. (pattern is `namespace:image-policy:tag` for tag, and  `namespace:image-policy` for the full image + tag).
 
 So basically, everytime a new `mastodon:nightly-XXXX-XX-XX` is pushed, my instance is automatically updated to this tag.

And this PR will allow me to do the same, directly with the main branch !

Why not using `:main` , `:latest` (or here, `:edge`) directly ?

Because the `:edge`  tag is not stable, without a proper fixed tag (never use `:branch-name` or `:latest` tag in production :smile:), you could end up with inconsistencies in your infra, eg the mastodon web container will be on `:edge` and the pulled container contains the code with the N state, and sidekiqs also using `:edge` could be in the N+3 commits state

I hope it's clearer now :smile:

**Comment by renchap** [comment]
We are already building `nightly` containers daily, and are considering stopping building a container for each merge on `main` as this takes a long time (which could be optimised, but is not trivial right now) and we are having a lot of activity/merges.

If we go this way, we will tag each nightly with `edge` so edge users will continue to get updates.

**Comment by forsamori** [comment]
@emilweth thanks for the comprehensive explanation! Yeah, that does help me understand the benefit - it also makes me realise that I have a lot to learn :D
